README.md
# Knapper

Simple, dependency-free knapsack solver

Master: [![Build Status: Master](https://travis-ci.com/es02/knapper.svg?branch=master)](https://travis-ci.com/es02/knapper)

Dev: [![Build Status: Dev](https://travis-ci.com/es02/knapper.svg?branch=dev)](https://travis-ci.com/es02/knapper)

## Getting Started

These instructions will get you a copy of the project installed in your composer dependencies.

### Prerequisites

PHP 7

### Installing

Adding knapper to your project is as simple as adding it to your composer.json

```
composer require es02/knapper
```

### Usage

```
include "../vendor/autoload.php";

use es02\knapper\Knapper;

$knapper = new Knapper($items, $boxes);
$results = $knapper->pack();
```

To use Knappper you will need an array of objects to be packed and an array of boxes to pack them into.
Optionally you can supply the following fields:

* maxCubic (Default: null) - Set an upper limit on cubic weight for boxes
* maxWeight (Default: null) - set an upper limit on physical weight for boxes
* weightType (Options: 'g', 'kg', 'oz', 'lb') (Default: 'g')  - Choose metric or imperial measurements for that weight

Some carriers require boxes in a shipment to remain below certain cubic/gross weight limits.
These options allow you to achieve this with minimal effort.

```
$knapper = new Knapper($items, $boxes, 15, 15 , 'kg');
$results = $knapper->pack();
```

Both arrays require the following fields:

* length
* width
* height
* weight

```
$items[] = array('length' => 10, 'width' => 50, 'height' => 200, 'weight' => 2);
$boxes[] = array('length' => 20, 'width' => 60, 'height' => 250, 'weight' => 2.5);
```

In addition both support the following optional fields:

* name (Default: null) - Allows you to set a human readable name for your item/box
* lengthType (Options: 'cm', 'm', 'in', 'ft') (Default: 'cm') - Sets measurement type
* weightType (Options: 'g', 'kg', 'oz', 'lb') (Default: 'g') - Sets measurement type

Boxes can have the following optional fields
* quantity  (Default: 0) - Sets whether or not a box size has a limited supply, leave at 0 for unlimited

Items can have the following optional fields
* thisWayUp  (Default: false) - Sets whether or not the item can be placed on it's side

*Where not manually set, optional fields will use their default values*

```
$items[] = array(
    'name' => "Sculpture"
    'length' => 10,
    'width' => 50,
    'height' => 200,
    'weight' => 2,
    'lengthType' => 'in',
    'weightType' => 'lb',
    'thisWayUp' => false
);
$boxes[] = array(
    'name' => "Large Packing Box",
    'length' => 20,
    'width' => 60,
    'height' => 250,
    'weight' => 2.5,
    'lengthType' => 'in',
    'weightType' => 'lb',
    'quantity' => 20
);
```

## Contributing

Please read [CONTRIBUTING.md]() for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/es02/knapper/tags).

## Authors

* **Aimi Hobson** - *Initial work* - [ArachnidsGrip](https://github.com/es02)

## License

This project is licensed under the GNU GPL - see the [LICENSE.md](LICENSE.md) file for details

## Acknowledgments

* ReadMe template from [PurpleBooth](https://gist.github.com/PurpleBooth/109311bb0361f32d87a2)
* Changelog Template from [KeepAChangelog](https://keepachangelog.com/en/1.0.0/)
* Contributing template from [PurpleBooth](https://gist.github.com/PurpleBooth/b24679402957c63ec426)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- Licence, Contributing guide and this Changelog
- Basic folder structure
- composer.json
- Base classes
- Models
- Controllers
- Added PHPUnit for dev builds only
- Added PHPLint for dev builds only
- Added PHP_CodeStandards for dev builds only
- Added TravisCI for Build Management
- Initial unit tests added

### Changed
- Readme to include actual information instead of base GitHub autogen file.
- Build status images to readme for master & dev branches
- Licence filename to match everything else, may change this back in future.

[Unreleased]: https://github.com/es02/knapper/tree/dev


example/index.php
<?php
namespace es02\example;

include "../vendor/autoload.php";

use es02\knapper\Knapper;

/**
 *
 * Obviously you would place items and boxes
 * in some kind of database and not hard code them like this
 *
 * Note: Default mesurement types are being used (Grams and Centimeters)
 *
 **/

$items = array(
    array(                  // Should be packed into Box1
        'name' =>'hats',
        'length' => 10,
        'width' => 50,
        'height' => 200,
        'weight' => 500
    ),
    array(                  // Should be packed into Box2
        'name' =>'bats',
        'length' => 100,
        'width' => 500,
        'height' => 150,
        'weight' => 1500
    ),
    array(                  // Should be packed into Box2
        'name' =>'cats',
        'length' => 150,
        'width' => 506,
        'height' => 205,
        'weight' => 2100
    ),array(                  // Should be packed into Box1
        'name' =>'rats',
        'length' => 100,
        'width' => 50,
        'height' => 20,
        'weight' => 800
    ),array(                  // Should be packed into Box1
        'name' =>'gnats',
        'length' => 10,
        'width' => 5,
        'height' => 20,
        'weight' => 200
    ),
);
$boxes = array(
    array(
        'name' =>'box1',
        'length' => 20,
        'width' => 60,
        'height' => 250,
        'weight' => 2500
    ),
    array(
        'name' =>'box2',
        'length' => 150,
        'width' => 600,
        'height' => 300,
        'weight' => 8000
    ),
    array(
        'name' =>'box3',
        'length' => 20,
        'width' => 60,
        'height' => 250,
        'weight' => 2500
    )
);

$knapper = new knapper($items, $boxes);
$results = $knapper->pack();

print_r($results);

?>
